server/src/main/java/org/elasticsearch/common/util/StringLiteralDeduplicator.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.common.util;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.util.Map;
+
+/**
+ * A cache in front of Java's string interning. This method assumes that it is only called with strings that are already part of the
+ * JVM's string pool so that interning them does not grow the pool. Calling it with strings not in the interned string pool is not
+ * advisable as its performance may deteriorate to slower than outright calls to {@link String#intern()}.
+ */
+public enum StringLiteralDeduplicator {
+    INSTANCE;
+
+    private static final Logger logger = LogManager.getLogger(StringLiteralDeduplicator.class);
+
+    private static final int MAX_SIZE = 1000;
```

Comment:
This will start thrashing once we exceed 1000 strings. I worry about the reuse of this `INSTANCE` for other purposes. Can we instead make a single instance that is used solely for the field capabilities `type` field by making a private static instance in the `IndexFieldCapabilities` class? If other needs arise for this deduplicator, those would thereby be separated completely.

Reply:
++ made it a private static cache for now just limited to this class